File: tests/test_ai_profile_update.py
#!/usr/bin/env python3
"""
AI隐式调用表单更新测试

测试场景：
1. AI从用户对话中提取结构化信息
2. AI调用ProfileForm.updateByAI()隐式更新用户画像
3. 验证数据是否正确保存到数据库

使用方法：
    cd D:\github.com\carrotmet\zyplusv2
    python tests\test_ai_profile_update.py

或者直接在浏览器控制台测试：
    // 模拟AI调用
    await ProfileForm.updateByAI(userId, {
        'career_path_preference': 'technical',
        'value_priorities': ['成就感', '创新']
    }, '从对话中提取');
"""

import sys
import os

# 添加backend到Python路径
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'backend'))

import json
from app.database import SessionLocal
from app.models_user_profile import UserProfile, UserProfileLog
from app.crud_user_profile import get_user_profile, update_user_profile
from app.schemas_user_profile import UserProfileUpdate


def test_ai_update_api():
    """
    测试后端API：POST /api/user-profiles/{user_id}/ai-update
    
    这是AI隐式调用的后端接口测试
    """
    print("=" * 60)
    print("测试1: AI更新API (后端)")
    print("=" * 60)
    
    db = SessionLocal()
    test_user_id = "test_ai_user_001"
    
    try:
        # 1. 创建测试用户画像（如果不存在）
        from app.crud_user_profile import get_or_create_user_profile
        profile = get_or_create_user_profile(db, test_user_id, "AI测试用户")
        print(f"✓ 测试用户创建/获取成功: {test_user_id}")
        print(f"  初始完整度: {profile.completeness_score}%")
        
        # 2. 模拟AI提取的数据
        ai_extracted_data = {
            "career_path_preference": "technical",  # 技术专家路径
            "value_priorities": ["成就感", "创新", "成长"],
            "ability_assessment": {
                "logic": 8,
                "creativity": 7,
                "communication": 6
            },
            "current_casve_stage": "analysis"  # 推进到分析阶段
        }
        
        print(f"\n✓ AI提取的数据:")
        for key, value in ai_extracted_data.items():
            print(f"  {key}: {value}")
        
        # 3. 调用更新函数（模拟API行为）
        updated_profile = update_user_profile(
            db=db,
            user_id=test_user_id,
            profile_update=UserProfileUpdate(**ai_extracted_data),
            update_type="ai_extraction"
        )
        
        print(f"\n✓ 更新成功!")
        print(f"  新完整度: {updated_profile.completeness_score}%")
        print(f"  职业路径: {updated_profile.career_path_preference}")
        print(f"  价值观: {updated_profile.value_priorities}")
        print(f"  CASVE阶段: {updated_profile.current_casve_stage}")
        
        # 4. 验证更新日志
        logs = db.query(UserProfileLog).filter(
            UserProfileLog.user_id == test_user_id
        ).order_by(UserProfileLog.timestamp.desc()).all()
        
        print(f"\n✓ 更新日志 ({len(logs)} 条):")
        for log in logs[:5]:
            print(f"  [{log.update_type}] {log.field_name}: {log.old_value} -> {log.new_value}")
        
        return True
        
    except Exception as e:
        print(f"✗ 测试失败: {e}")
        import traceback
        traceback.print_exc()
        return False
    finally:
        db.close()


def test_rag_service_integration():
    """
    测试RAG服务集成AI更新的完整流程
    
    模拟：用户输入 -> RAG处理 -> 提取信息 -> AI更新画像
    """
    print("\n" + "=" * 60)
    print("测试2: RAG服务集成AI更新")
    print("=" * 60)
    
    db = SessionLocal()
    test_user_id = "test_ai_user_002"
    
    try:
        # 1. 准备测试数据
        from app.crud_user_profile import get_or_create_user_profile
        profile = get_or_create_user_profile(db, test_user_id, "RAG集成测试用户")
        
        user_message = "我喜欢编程，想走技术路线，擅长逻辑思维，重视创新和成就感"
        
        print(f"✓ 测试用户: {test_user_id}")
        print(f"✓ 用户输入: {user_message}")
        
        # 2. 调用RAG服务处理消息
        from app.services.rag_service import get_rag_service
        rag_service = get_rag_service()
        
        profile_dict = {
            "holland_code": profile.holland_code,
            "mbti_type": profile.mbti_type,
            "value_priorities": profile.value_priorities,
            "ability_assessment": profile.ability_assessment,
            "career_path_preference": profile.career_path_preference,
            "current_casve_stage": profile.current_casve_stage,
            "completeness_score": profile.completeness_score
        }
        
        result = rag_service.process_message(
            user_message=user_message,
            user_profile=profile_dict,
            conversation_history=[]
        )
        
        print(f"\n✓ RAG处理结果:")
        print(f"  意图: {result.get('intent')}")
        print(f"  回复: {result.get('reply', '')[:50]}...")
        print(f"  提取信息: {result.get('extracted_info')}")
        print(f"  画像更新建议: {result.get('profile_updates')}")
        
        # 3. 应用提取的更新（模拟后端API行为）
        profile_updates = result.get('profile_updates', {})
        if profile_updates:
            # 过滤有效字段
            valid_fields = [
                'holland_code', 'mbti_type', 'value_priorities', 'ability_assessment',
                'career_path_preference', 'current_casve_stage', 'universal_skills', 'resilience_score'
            ]
            update_data = {k: v for k, v in profile_updates.items() if k in valid_fields and v}
            
            if update_data:
                updated_profile = update_user_profile(
                    db=db,
                    user_id=test_user_id,
                    profile_update=UserProfileUpdate(**update_data),
                    update_type="ai_extraction"
                )
                
                print(f"\n✓ AI自动更新成功!")
                print(f"  更新字段: {list(update_data.keys())}")
                print(f"  新完整度: {updated_profile.completeness_score}%")
            else:
                print(f"\n⚠ RAG未提取到有效更新字段")
        else:
            print(f"\n⚠ RAG未返回画像更新建议")
        
        return True
        
    except Exception as e:
        print(f"✗ 测试失败: {e}")
        import traceback
        traceback.print_exc()
        return False
    finally:
        db.close()


def generate_frontend_test_code():
    """
    生成前端测试代码（可在浏览器控制台执行）
    """
    print("\n" + "=" * 60)
    print("测试3: 前端AI隐式调用代码示例")
    print("=" * 60)
    
    code = '''
// ==================== 在浏览器控制台执行 ====================

// 1. 获取当前用户ID
const userId = localStorage.getItem('user_id');
console.log('当前用户ID:', userId);

// 2. 模拟AI从对话中提取信息并隐式更新
async function testAIProfileUpdate() {
    console.log('开始测试AI隐式更新...');
    
    // 模拟AI提取的结构化数据
    const aiExtractedData = {
        career_path_preference: 'technical',  // 技术专家路径
        value_priorities: ['成就感', '创新', '成长'],
        ability_assessment: {
            logic: 8,
            creativity: 7,
            communication: 6
        },
        current_casve_stage: 'analysis',  // 推进到分析阶段
        resilience_score: 8
    };
    
    console.log('AI提取的数据:', aiExtractedData);
    
    // 调用ProfileForm.updateByAI进行隐式更新
    // 注意：这需要ProfileForm模块已加载
    if (typeof ProfileForm !== 'undefined') {
        const result = await ProfileForm.updateByAI(
            userId,
            aiExtractedData,
            '从对话"我想成为技术专家"中提取'
        );
        
        console.log('AI更新结果:', result);
        
        if (result.success) {
            console.log('✓ AI隐式更新成功!');
            console.log('  更新字段:', result.data.updated_fields);
            
            // 刷新UI显示
            if (typeof updateUIFromProfileData === 'function') {
                updateUIFromProfileData(result.data.profile);
            }
        } else {
            console.error('✗ AI更新失败:', result.error);
        }
    } else {
        console.error('ProfileForm模块未加载');
    }
}

// 执行测试
testAIProfileUpdate();

// ==================== 测试结束 ====================
'''
    print(code)
    
    # 保存到文件
    js_file = os.path.join(os.path.dirname(__file__), 'test_ai_profile_update.js')
    with open(js_file, 'w', encoding='utf-8') as f:
        f.write(code)
    print(f"\n✓ 前端测试代码已保存到: {js_file}")


def main():
    """运行所有测试"""
    print("\n" + "=" * 60)
    print("AI隐式调用表单更新测试套件")
    print("=" * 60)
    
    results = []
    
    # 测试1: AI更新API
    results.append(("AI更新API", test_ai_update_api()))
    
    # 测试2: RAG服务集成
    results.append(("RAG服务集成", test_rag_service_integration()))
    
    # 测试3: 生成前端测试代码
    generate_frontend_test_code()
    results.append(("前端测试代码生成", True))
    
    # 总结
    print("\n" + "=" * 60)
    print("测试结果总结")
    print("=" * 60)
    for name, passed in results:
        status = "✓ 通过" if passed else "✗ 失败"
        print(f"  {name}: {status}")
    
    all_passed = all(r[1] for r in results)
    print("\n" + ("✓ 所有测试通过!" if all_passed else "✗ 部分测试失败"))
    
    return all_passed


if __name__ == "__main__":
    success = main()
    sys.exit(0 if success else 1)

File: tests/test_ai_profile_update.js
/**
 * AI隐式调用表单更新测试 - 前端代码
 * 
 * 在浏览器控制台执行以下代码测试AI隐式更新功能
 */

// ==================== 测试1: 基础AI更新 ====================

/**
 * 测试AI隐式更新画像
 * 
 * 使用场景：AI从用户对话中提取信息后，自动更新用户画像
 * 调用方式：ProfileForm.updateByAI(userId, updates, reason)
 */
async function testAIProfileUpdate() {
    console.log('=== 测试AI隐式更新画像 ===');
    
    // 获取当前用户ID
    const userId = localStorage.getItem('user_id');
    if (!userId) {
        console.error('错误：未找到用户ID，请先登录');
        return;
    }
    console.log('当前用户ID:', userId);
    
    // 模拟AI从对话中提取的结构化数据
    // 示例对话："我喜欢编程，想走技术路线，擅长逻辑思维，重视创新和成就感"
    const aiExtractedData = {
        career_path_preference: 'technical',  // 技术专家路径
        value_priorities: ['成就感', '创新', '成长'],
        ability_assessment: {
            logic: 8,
            creativity: 7,
            communication: 6
        },
        current_casve_stage: 'analysis',  // 推进到分析阶段
        resilience_score: 8
    };
    
    console.log('AI提取的数据:', aiExtractedData);
    
    // 检查ProfileForm模块是否可用
    if (typeof ProfileForm === 'undefined') {
        console.error('错误：ProfileForm模块未加载，请确保已引入profile-form.js');
        return;
    }
    
    try {
        // 调用AI更新接口（隐式更新，不打开表单）
        const result = await ProfileForm.updateByAI(
            userId,
            aiExtractedData,
            '从对话"我想成为技术专家"中提取'
        );
        
        console.log('AI更新结果:', result);
        
        if (result.success) {
            console.log('✓ AI隐式更新成功!');
            console.log('  更新字段:', result.data.updated_fields);
            console.log('  新完整度:', result.data.profile.completeness_score + '%');
            
            // 刷新UI显示
            if (typeof updateUIFromProfileData === 'function') {
                updateUIFromProfileData(result.data.profile);
                console.log('✓ UI已刷新');
            }
        } else {
            console.error('✗ AI更新失败:', result.error);
        }
    } catch (error) {
        console.error('✗ 测试出错:', error);
    }
}


// ==================== 测试2: 模拟RAG对话流程 ====================

/**
 * 模拟完整的RAG对话处理流程
 * 
 * 流程：用户输入 -> RAG处理 -> 提取信息 -> AI更新 -> 更新UI
 */
async function testRAGFlowWithAIUpdate() {
    console.log('=== 测试RAG对话流程（含AI更新）===');
    
    const userId = localStorage.getItem('user_id');
    if (!userId) {
        console.error('错误：未找到用户ID');
        return;
    }
    
    // 模拟用户消息
    const userMessage = "我擅长逻辑思维，喜欢解决复杂问题，希望走技术专家路线";
    console.log('用户输入:', userMessage);
    
    // 模拟RAG服务返回的结果
    const mockRAGResult = {
        reply: "逻辑思维和技术专家路线很匹配！你有很强的分析能力。",
        intent: "career_advice",
        extracted_info: [
            { field: "ability_assessment", value: { logic: 8, analysis: 9 }, confidence: 0.9 },
            { field: "career_path_preference", value: "technical", confidence: 0.95 }
        ],
        profile_updates: {
            career_path_preference: "technical",
            ability_assessment: { logic: 8, analysis: 9 }
        }
    };
    
    console.log('RAG处理结果:', mockRAGResult);
    
    // 调用AI更新
    if (mockRAGResult.profile_updates && Object.keys(mockRAGResult.profile_updates).length > 0) {
        const result = await ProfileForm.updateByAI(
            userId,
            mockRAGResult.profile_updates,
            'RAG从对话中提取'
        );
        
        if (result.success) {
            console.log('✓ RAG流程完成：对话 -> 提取 -> AI更新 -> UI刷新');
            
            // 显示通知
            if (typeof showExtractedInfoNotification === 'function') {
                showExtractedInfoNotification(mockRAGResult.extracted_info);
            }
        }
    }
}


// ==================== 测试3: 批量字段更新 ====================

/**
 * 测试同时更新多个字段
 */
async function testBatchAIUpdate() {
    console.log('=== 测试批量AI更新 ===');
    
    const userId = localStorage.getItem('user_id');
    
    // 模拟一次完整的画像信息收集
    const completeProfileData = {
        holland_code: 'RIA',
        mbti_type: 'INTJ',
        value_priorities: ['成就感', '创新', '成长', '自由'],
        ability_assessment: {
            logic: 9,
            creativity: 7,
            communication: 6,
            teamwork: 5,
            leadership: 7,
            analysis: 9
        },
        career_path_preference: 'technical',
        universal_skills: {
            communication: 7,
            teamwork: 6,
            critical_thinking: 9,
            creativity: 7,
            problem_solving: 9,
            adaptability: 8
        },
        resilience_score: 8,
        current_casve_stage: 'analysis'
    };
    
    console.log('批量更新数据:', completeProfileData);
    
    const result = await ProfileForm.updateByAI(
        userId,
        completeProfileData,
        '完整画像信息收集'
    );
    
    if (result.success) {
        console.log('✓ 批量更新成功!');
        console.log('  更新字段数:', result.data.updated_fields.length);
        console.log('  预计完整度:', result.data.profile.completeness_score + '%');
    }
}


// ==================== 测试4: 对比手动表单和AI隐式调用 ====================

/**
 * 对比两种更新方式
 */
async function compareUpdateMethods() {
    console.log('=== 对比更新方式 ===');
    
    const userId = localStorage.getItem('user_id');
    
    console.log('方式1: 手动表单（用户主动）');
    console.log('  - 调用: ProfileForm.open(userId, data, callback)');
    console.log('  - 特点: 弹出表单，用户可见，可编辑');
    console.log('  - 适用: 用户主动完善信息');
    
    console.log('\n方式2: AI隐式调用（AI自动）');
    console.log('  - 调用: ProfileForm.updateByAI(userId, updates, reason)');
    console.log('  - 特点: 静默更新，无弹窗，自动完成');
    console.log('  - 适用: AI从对话中提取信息');
    
    console.log('\n两种方式的API端点：');
    console.log('  - 手动表单: POST /api/user-profiles/{user_id}/form-update');
    console.log('  - AI隐式:   POST /api/user-profiles/{user_id}/ai-update');
}


// ==================== 快速测试入口 ====================

// 在控制台输入以下命令执行测试：
// testAIProfileUpdate()      - 测试基础AI更新
// testRAGFlowWithAIUpdate()  - 测试RAG完整流程
// testBatchAIUpdate()        - 测试批量更新
// compareUpdateMethods()     - 对比更新方式

console.log('AI隐式调用测试脚本已加载');
console.log('可用测试函数:');
console.log('  testAIProfileUpdate() - 基础AI更新测试');
console.log('  testRAGFlowWithAIUpdate() - RAG流程测试');
console.log('  testBatchAIUpdate() - 批量更新测试');
console.log('  compareUpdateMethods() - 对比更新方式');
